x-pack/filebeat/input/cel/integration_test.go

```diff
@@ -0,0 +1,521 @@
+	"time"
+
+	"google.golang.org/protobuf/types/known/structpb"
+
+	"github.com/elastic/beats/v7/libbeat/tests/integration"
+	filebeat "github.com/elastic/beats/v7/x-pack/filebeat/cmd"
+	"github.com/elastic/elastic-agent-client/v7/pkg/client/mock"
+	"github.com/elastic/elastic-agent-client/v7/pkg/proto"
+)
+
+// TestCheckinV2 is an integration test that checks that CEL input reports in
+// the expected statuses. Specifically it configures a filebeat instance to
+// run a CEL input with two streams as well as it monitors the reported state
+// by spawning an elastic-agent V2 mock server.
+// This test also spawns two http servers for making the CEL input streams
+// to report different states that are checked to match the expected states.
+func TestCheckinV2(t *testing.T) {
+	// make sure there is an ES instance running
+	integration.EnsureESIsRunning(t)
+	esConnectionDetails := integration.GetESURL(t, "http")
+	outputHosts := []interface{}{fmt.Sprintf("%s://%s:%s", esConnectionDetails.Scheme, esConnectionDetails.Hostname(), esConnectionDetails.Port())}
+	outputUsername := esConnectionDetails.User.Username()
+	outputPassword, _ := esConnectionDetails.User.Password()
+	outputProtocol := esConnectionDetails.Scheme
+
+	invalidResponse := []byte("invalid json")
+	validResponse := []byte("{\"ip\":\"0.0.0.0\"}")
+
+	// http server for the first CEL input stream
+	serverOneResponse := validResponse
+	svrOne := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		_, _ = w.Write(serverOneResponse)
+	}))
+	defer svrOne.Close()
+
+	// http server for the second CEL input stream
+	serverTwoResponse := validResponse
+	svrTwo := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		_, _ = w.Write(serverTwoResponse)
+	}))
+	defer svrTwo.Close()
+
+	// allStreams is an elastic-agent configuration with an ES output and one CEL
+	// input with two streams.
+	allStreams := []*proto.UnitExpected{
+		{
+			Id:             "output-unit",
+			Type:           proto.UnitType_OUTPUT,
+			ConfigStateIdx: 0,
+			State:          proto.State_HEALTHY,
+			LogLevel:       proto.UnitLogLevel_INFO,
+			Config: &proto.UnitExpectedConfig{
+				Id:   "default",
+				Type: "elasticsearch",
+				Name: "elasticsearch",
+				Source: integration.RequireNewStruct(t, map[string]interface{}{
+					"type":                  "elasticsearch",
+					"hosts":                 outputHosts,
+					"username":              outputUsername,
+					"password":              outputPassword,
+					"protocol":              outputProtocol,
+					"enabled":               true,
+					"ssl.verification_mode": "none",
+				}),
+			},
+		},
+		{
+			Id:             "input-unit-1",
+			Type:           proto.UnitType_INPUT,
+			ConfigStateIdx: 0,
+			State:          proto.State_HEALTHY,
+			LogLevel:       proto.UnitLogLevel_DEBUG,
+			Config: &proto.UnitExpectedConfig{
+				Id:   "cel-cel-1e8b33de-d54a-45cd-90da-23ed71c482e5",
+				Type: "cel",
```

Comment:
I tend to believe that helper functions in tests reduce readability because you have to jump around in so many different places to essentially read the same amount of code or even worse if you make your helper functions too smart now you have to jump all around but also account for parameters that change values between calls. So I won't proceed with your proposal here. Feel free to un-resolve if you have strong feelings

Reply:
I agree with @pkoutsovasilis here despite my general wishes to use helpers tests to reduce incidental text. I spent a fair bit of time trying to factor this out to get this to be a helper, but it's fundamentally a closure, and parameterising all the values that are closed over to make it a helper would be counterproductive IMO.